### PR TITLE
HARP-5401: Add option to set maximum frame rate

### PR DESCRIPTION
**Pull request by StefanDachwitz**
Signed-off-by: Stefan Dachwitz <StefanDachwitz@users.noreply.github.com>

### CODE REVIEW

File: @here/harp-mapview/lib/MapView.ts [hunk 1]
```diff
@@ -1759,16 +1784,35 @@ export class MapView extends THREE.EventDispatcher {
 
         this.m_pixelToWorld = undefined;
 
-        const cameraPitch = Math.min(
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: I'm guessing this was a mistake? It seems not related to your commit.

File: @here/harp-mapview/lib/MapView.ts [hunk 2]
```diff
@@ -1759,16 +1784,35 @@ export class MapView extends THREE.EventDispatcher {
...
     }
 
+    private detectCurrentFps(now: number) {
+        if (this.m_previousRequestAnimationTime === undefined) {
+            this.m_previousRequestAnimationTime = now;
+            return;
+        }
+
+        // Skip the first frames, they are from not originated from requestAnimationFrame()
+        if (this.m_frameNumber > 5 /*&& now - this.m_previousRequestAnimationTime > 5*/) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: should this comment stay here? Why?

File: @here/harp-mapview/lib/MapView.ts [hunk 3]
```diff
@@ -1780,8 +1824,58 @@ export class MapView extends THREE.EventDispatcher {
...
+
+        const vSyncFrameTime = 1000 / (this.m_detectedFps !== undefined ? this.m_detectedFps : 30);
+        const frameInterval = 1000 / this.m_maxFps;
+
+        const startTime =
+            this.m_previousFrameTimeStamp === undefined
+                ? performance.now()
+                : this.m_previousFrameTimeStamp;
+
+        // Magic ingredient to compensate time flux.
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: this is 3ms? I think it should be added to the variable name

File: @here/harp-mapview/lib/MapView.ts [hunk 4]
```diff
@@ -1759,16 +1784,35 @@ export class MapView extends THREE.EventDispatcher {
...
-        const cameraPitch = Math.min(
-            MapViewUtils.extractYawPitchRoll(this.camera.quaternion).pitch,
-            Math.PI / 3
-        );
+        const cameraPitch = MapViewUtils.extractYawPitchRoll(this.camera.quaternion).pitch;
 
-        const distance = Math.abs(this.camera.position.z) / Math.cos(cameraPitch);
+        const distance = Math.min(
+            Math.abs(this.camera.position.z) / Math.cos(cameraPitch),
+            this.camera.position.z * 2
```
[the remark this answers is not in the document]
Comment (StefanDachwitz) [reply @here/harp-mapview/lib/MapView.ts]: That seems to be a merge issue. Reverting this.

File: @here/harp-mapview/lib/MapView.ts [hunk 5]
```diff
@@ -1791,6 +1839,27 @@ export class MapView extends THREE.EventDispatcher {
...
+        }
+
+        // Skip the first frames, they are from not originated from requestAnimationFrame()
+        if (this.m_frameNumber > 5) {
+            const currentFps = 1000 / (now - this.m_previousRequestAnimationTime);
+            this.m_frameTimeRing[this.m_frameTimeIndex] = currentFps;
+            this.m_frameTimeIndex = (this.m_frameTimeIndex + 1) % FRAME_RATE_RING_SIZE;
+
+            let sum: number = 0;
+            for (let i = 0; i < FRAME_RATE_RING_SIZE; i++) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: minor, you could save some code by using reduce(): https://www.w3schools.com/jsref/jsref_reduce.asp

File: @here/harp-mapview/lib/MapView.ts [hunk 6]
```diff
@@ -952,6 +989,17 @@ export class MapView extends THREE.EventDispatcher {
         this.m_forceCameraAspect = aspect;
     }
 
+    /**
+     * Maximum FPS. If defined (and > 0) it is the maximum FPS that is used.
+     */
+    set maxFps(fps: number | undefined) {
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: why can the user set an undefined maxFps?

File: @here/harp-mapview/lib/MapView.ts [hunk 7]
```diff
@@ -1791,6 +1839,27 @@ export class MapView extends THREE.EventDispatcher {
...
+        if (this.m_frameNumber > 5) {
+            const currentFps = 1000 / (now - this.m_previousRequestAnimationTime);
+            this.m_frameTimeRing[this.m_frameTimeIndex] = currentFps;
+            this.m_frameTimeIndex = (this.m_frameTimeIndex + 1) % FRAME_RATE_RING_SIZE;
+
+            let sum: number = 0;
+            for (let i = 0; i < FRAME_RATE_RING_SIZE; i++) {
+                sum += this.m_frameTimeRing[i];
+            }
+            this.m_detectedFps = Math.floor(sum / FRAME_RATE_RING_SIZE);
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: this is slightly wrong, if the ring is only partially filled, then it should divide by the number of valid entries (?)

File: @here/harp-mapview/lib/MapView.ts [hunk 8]
```diff
@@ -1802,8 +1871,61 @@ export class MapView extends THREE.EventDispatcher {
...
+        if (this.m_targetRequestAnimationTime === undefined) {
+            return;
+        }
+
+        if (this.m_animationFrameHandle !== undefined) {
+            cancelAnimationFrame(this.m_animationFrameHandle);
+            this.m_animationFrameHandle = undefined;
+        }
+
+        this.detectCurrentFps(now);
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: is this correct? This will detect the fps using a postponed animation frame, which will artificially inflate the FPS

Comment (okanishchev) [reply @here/harp-mapview/lib/MapView.ts]: in this case we need detected fps only to measure delay to next rAF callback,
and this delay is computed by browser using previous callbacks duration.
As there is no standard algo for this delay, and no api method, we need to estimate it somehow.

Also, we can't do rAF, then setTimeout (or similar) to next rAF - that won't work as expected, especially on high cpu load.

File: @here/harp-mapview/lib/MapView.ts [hunk 9]
```diff
@@ -965,6 +998,17 @@ export class MapView extends THREE.EventDispatcher {
...
 
+    /**
+     * Maximum FPS. If defined (and > 0) it is the maximum FPS that is used.
+     */
+    set maxFps(fps: number) {
+        this.m_maxFps = Math.max(0, fps);
+    }
+
+    get maxFps(): number {
+        return this.m_maxFps > 0 ? this.m_maxFps : 0;
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: Math.max(0, this.m_maxFps) would be equivalent?

Comment (okanishchev) [reply @here/harp-mapview/lib/MapView.ts]: Yeah, I think so. As we already using max everywhere, we can also do it here.
